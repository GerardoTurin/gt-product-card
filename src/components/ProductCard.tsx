import { createContext } from 'react';
import useProduct from '../hooks/useProduct';
import { InitialValueProps, Product, ProductCardHandlerProps, ProductContextProps, onChangeArgs } from '../interfaces/prodInterfaces';
import styles from '../styles/styles.module.css';
import React from "react";


export interface ProductCardProps {
    product: Product;
    //children?: ReactElement | ReactElement[];    // Se le dice que puede recibir un elemento o varios elementos de React
    children: ( args: ProductCardHandlerProps ) => JSX.Element;
    className?: string;
    style?: React.CSSProperties;
    onChange?: ( args: onChangeArgs) => void;
    value?: number;
    initialValue?: InitialValueProps;
}


export const ProductContext = createContext( {} as ProductContextProps );   // Se crea el contexto con un valor inicial vacio
const { Provider } = ProductContext;    // Se extrae el provider del contexto



const ProductCard = ({ children, product, className, style, onChange, value, initialValue }: ProductCardProps) => {
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValue });


    return (
        <Provider value={{ counter, increaseBy, product, maxCount }}>
            <div className={ `${ styles.productCard } ${ className }` } style={ style }>
                { children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValue?.maxCount,
                    product,
                    increaseBy,
                    reset
                
                }) }
            </div>
        </Provider>
    )
};


export default ProductCard;