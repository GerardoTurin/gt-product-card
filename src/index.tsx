import * as React from 'react';
export * from './components';

// Delete me
export const Thing = () => {
  return <div>the snozzberries taste like snozzberries</div>;
};
