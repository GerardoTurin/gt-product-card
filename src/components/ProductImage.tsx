import { useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImg from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import React from "react";



export interface ProductImageProps {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}


export const ProductImage = ({ img, className, style }: ProductImageProps) => {
    const { product } = useContext( ProductContext );   // Se extrae el producto del contexto
    let imgToShow: string;

    if ( img ) {
        imgToShow = img;
    } else if ( product.img ) {
        imgToShow = product.img;
    } else {
        imgToShow = noImg;
    }


    return (
        <img 
            className={ `${ styles.productImg } ${ className }` } style={ style } 
            alt="Product" src={ imgToShow } />
    )
};