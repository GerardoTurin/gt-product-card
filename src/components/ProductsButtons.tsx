import { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import React from "react";

export interface ProductsButtonsProps {
    className?: string;
    style?: React.CSSProperties;
}



export const ProductsButtons = ({ className, style }: ProductsButtonsProps) => {
    const { counter, increaseBy, maxCount } = useContext( ProductContext );   // Se extraen las funciones del contexto

    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount, 
        [counter, maxCount]    
    );

    

    return (
        <div className={ `${ styles.buttonsContainer } ${ className }` } style={ style }>
            <button className={ styles.buttonMinus } onClick={ () => increaseBy( -1 ) }>
                -
            </button>

            <div className={ styles.countLabel }>
                { counter }
            </div>

            <button 
                className={ `${ styles.buttonAdd } ${ isMaxReached() && styles.disabled }` }
                onClick={ () => increaseBy( +1 ) }>
                +
            </button>
        </div>
    )
};