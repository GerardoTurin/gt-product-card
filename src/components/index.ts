//export { default as ProductCard } from './ProductCard';
export * from './ProductImage';
export * from './ProductTitle';
export * from './ProductsButtons';


import { ProductCardHOCProps } from '../interfaces/prodInterfaces';
import  ProductCardHOC  from './ProductCard';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductsButtons } from './ProductsButtons';

const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductsButtons
});

export default ProductCard;