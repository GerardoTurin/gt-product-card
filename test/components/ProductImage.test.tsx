import React from "react";
import renderer from "react-test-renderer";
import { product2 } from '../data/products';
import ProductCard from "../../src/components/ProductCard";
import { ProductImage } from "../../src/components";



describe("ProductImage", () => {
    test("Debe mostrar el componente con la imagen personalizado", () => {
        const wrapper = renderer.create(<ProductImage img="https://via.placeholder.com/150" />);

        //console.log(wrapper.toJSON());
        expect(wrapper.toJSON()).toMatchSnapshot();
    });


    test("Debe mostrar el componente con la imagen del producto", () => {
        const wrapper = renderer.create(
            <ProductCard product={ product2 }>
                {
                    () => (
                        <>
                            <ProductImage />
                        </>
                    )
                }
            </ProductCard>
        );
        
        //console.log(wrapper.toJSON());
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});