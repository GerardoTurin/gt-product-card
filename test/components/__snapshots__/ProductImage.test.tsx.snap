// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./no-Image.jpg"
  />
</div>
`;

exports[`ProductImage Debe mostrar el componente con la imagen personalizado 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="https://via.placeholder.com/150"
/>
`;
